server/src/main/java/io/crate/expression/scalar/AgeFunction.java

```diff
@@ -0,0 +1,226 @@
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.StringType;
+import org.apache.commons.math3.util.Pair;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+
+/**
+ * This class implements the pg_catalog.age function
+ * Usage example:
+ * <pre>
+ *    SELECT pg_catalog.age(timestamp '2021-11-06')
+ *
+ * </pre>
+ *
+ */
+public class AgeFunction extends Scalar<Long, Object>
+{
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "age");
+    private final Signature signature;
+    private final Signature boundSignature;
+
+
+    /**
+     * Registers the pg_catalog function in the {@link ScalarFunctionModule}.
+     * It takes as input a TIMESTAMP  or two TIMESTAMP and optionally TIMEZONE
+     * and produces a LONG
+     * @param module the {@link ScalarFunctionModule}
+     */
+    public static void register(ScalarFunctionModule module)
+    {
+        List<DataType<?>> supportedTimestampTypes = List.of(
+            DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMP, DataTypes.LONG);
+        for (DataType<?> dataType : supportedTimestampTypes)
+        {
+
+            module.register(
+                Signature.scalar(
+                    NAME,
+                    dataType.getTypeSignature(),
+                    DataTypes.LONG.getTypeSignature()
+                ), AgeFunction::new
+            );
+            module.register(
+                Signature.scalar(
+                    NAME,
+                    DataTypes.STRING.getTypeSignature(),
+                    dataType.getTypeSignature(),
+                    DataTypes.LONG.getTypeSignature()
+                ), AgeFunction::new
+            );
+
+            module.register(
+                Signature.scalar(
+                    NAME,
+                    dataType.getTypeSignature(),
+                    dataType.getTypeSignature(),
+                    DataTypes.LONG.getTypeSignature()
+                ),
+                AgeFunction::new
+            );
+            module.register(
+                Signature.scalar(
+                    NAME,
+                    DataTypes.STRING.getTypeSignature(),
+                    dataType.getTypeSignature(),
+                    dataType.getTypeSignature(),
+                    DataTypes.LONG.getTypeSignature()
+                ),
+                AgeFunction::new
+            );
+
+        }
+    }
+
+
+    public AgeFunction(Signature signature, Signature boundSignature)
+    {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+
+    @Override
+    public Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>... args)
+    {
+
+        assert args.length >= 1 &&  args.length <= 4 :
```

Comment:
I think without supporting timezones we can simplify roughly like:
pseudocode
```
if (args.len == 1) {
    long millis = convertArgToMillis(args[0]) //use implicitCast of TimestampType
    return millis % 86400000; //see curdate, we need to return t - curdate
} else {
  return convertArgToMillis(args[1]) - convertArgToMillis(args[0])
}
```

Reply:
Actually, we can even replace `if ` part by `UnaryScalar` and `else` part by `BinaryScalar`